server/graphql-tx-protocol.md

```diff
@@ -0,0 +1,171 @@
+# GraphQL Transactions over WebSocket Protocol
+
+Running GraphQL queries and mutations in a transaction over a WebSocket connection.
+
+## Connection
+
+The subprotocol name is `graphql-tx`.
+
+```http
+GET /v1/graphql HTTP/1.1
+Host: localhost:8080
+Upgrade: websocket
+Connection: Upgrade
+Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==
+Sec-WebSocket-Version: 13
+Sec-WebSocket-Protocol: graphql-tx
+```
+
+## Client - Server communication
+
+All messages between client and server are [JSON](https://www.json.org) objects with a
+compulsory `type` key and an optional `payload` key. The `type` represents the kind of message and
+the `payload` has relevant data. Few messages from server has `id` key which is a unique WebSocket
+id and `request_id` for query responses.
+
+#### 1. Init
+
+The `init` message is sent to initialise the transaction. It is identical in behaviour of PostgreSQL `BEGIN` command.
+A Postgres transaction is initiated. You can set the session variables using `headers` key and the [transaction isolation level](https://www.postgresql.org/docs/current/transaction-iso.html)
+using `isolation` key in the `payload`. Only one transaction is allowed in a single WebSocket connection.
+
+- `type: init`
+- `payload:`
+    * `isolation?: any of 'read-committed'(default), 'serializable' and 'repeatable-read'`
+    * `headers?: JSON object with String values`
+
+
+Client Message:-
+```json
+{
+  "type": "init",
+  "payload": {
+    "isolation": "serializable",
+    "headers": {
+      "X-Hasura-Role": "user",
+      "X-Hasura-User-Id": "1"
+    }
+  }
+}
+```
+Server Message on Success:-
+```json
+{
+  "type": "initialised"
+}
+```
+Server Message on Error:-
+```json
+{
+  "type": "init_error",
+  "payload": "transaction cannot be initialised more than once in a single WebSocket session"
+}
+```
+
+#### 2. Execute
+
+The `execute` message is sent to execute a GraphQL query or mutation. Mutations executed will not be committed in the database
+until `commit` message is sent. The `request_id` in `payload` is used to identify the responses from the server for the queries
+sent by the client.
+
+- `type: execute`
+- `payload: Object`
+    * `request_id?: String`
+    * `query: GraphQL query object`
```

Comment:
What do you think about calling this `operation` instead of `query`? Otherwise this leads to query within query object which is quite error-prone. E.g.

```
           ws.send(JSON.stringify({
                'type': 'execute',
                'payload': {
                    'request_id': 'some-unique-id',
                    'query': {
                        'query': 'query u { users { id name } } ',
                        'variables': {}
                    }
                }
            }))
```